Benchmarking

exfletch opened the pull request:
Let me know if you have any thoughts on the process here!

amakarowsky commented on validation/build_validation_databases.R:
```diff
@@ -0,0 +1,63 @@
+output_database <- "ref" # "test"
+required_file_names <- c("ref_circuit_details.csv", "ref_meta_data.json", "ref_raw_data.csv", "ref_site_details.csv")
+required_files <- list()
+# for (fn in required_file_names) {c(required_files, paste(file_prefix, fn, sep=""))}
```
Dead line

amakarowsky commented on validation/validate_results.R:
```diff
@@ -0,0 +1,160 @@
+# Validate benchmarking dataset results
+# Step 1: Ensure you have read the README in this directory and followed the instructions on creating the reference
+#         and test data
+# Step 2: Set tool_directory to the root directory of the DER disturbance analysis repo in your file system
```
Is it possible to get tool_directory automatically? Potentially something along the lines of [https://stackoverflow.com/questions/1815606/determine-path-of-the-executing-script](url) ?

exfletch replied:
It's a bit nasty, I've updated it to use the root directory by default with the option to override it if it detects that it's running somewhere else

amakarowsky reviewed (commented):
Looks good to me Jack, just one typo line.

The ability to automatically fetch different root directories of the repo on user-varying preceding file paths would be handy. Not sure how flexible this is in R however

phoebeheywood commented on validation/build_validation_databases.R:
```diff
@@ -0,0 +1,63 @@
+# To use this file, run it as an R script in the validation directory
```
Suggestion that things should be run from the tool directory since it work best with RStudio and most paths in other scripts are relative to the tool directory

phoebeheywood reviewed (commented):
I had a few problems running the build_validation_databases.R script which we should run through before I review any further

exfletch commented on validation/build_validation_databases.R:
```diff
@@ -0,0 +1,63 @@
+
+tool_directory <- "~/UNSW/MATCH/DER_disturbance_analysis"
+source(sprintf("%s/BDInterface/interface.R", tool_directory))
+library("logging")
```
- [ ] add logging library to install file

exfletch commented on validation/build_validation_databases.R:
```diff
@@ -0,0 +1,63 @@
+if (length(data_dirs) > 0){
+    for (dir in data_dirs){
+        all_files_in_dir <- list.files(dir)
+        required_files_in_dir <- required_files %in% all_files_in_dir
```
replaced references to `required_files` with `required_file_names`

phoebeheywood commented on validation/validate_results.R:
```diff
@@ -0,0 +1,160 @@
+            # check databases
+            ref_db_con <- RSQLite::dbConnect(RSQLite::SQLite(), sprintf("%s/%s", dir, ref_db_name))
+            test_db_con <- RSQLite::dbConnect(RSQLite::SQLite(), sprintf("%s/%s", dir, test_db_name))
+            compare_dbs(ref_db_con, test_db_con, TRUE, dir)
```
Check dbs exist before comparing

phoebeheywood commented on validation/validate_results.R:
```diff
@@ -0,0 +1,160 @@
+}
+
+
+if (length(data_dirs) > 0){
```
Maybe set this to log to console if files not found?

phoebeheywood commented on validation/validate_results.R:
```diff
@@ -0,0 +1,160 @@
+            compare_dfs(ref_underlying_data, test_underlying_data, event_name=dir)
+        }
+    }
+}
```
Currently nothing is logged if there are no differences. It would be worth printing something so that the user knows it ran successfully

exfletch replied:
Good thinking

phoebeheywood commented on validation/validate_results.R:
```diff
@@ -0,0 +1,160 @@
+    column_diff <- difference_between_lists(ref_columns, test_columns)
+
+    if (length(column_diff$reference_only) > 0) {
+        logging::loginfo(
```
It would be useful to also see which dfs are being compared when looking at the log

exfletch replied:
I'm now printing which CSV/dfs are being testing when this function is triggered

exfletch commented on validation/build_validation_databases.R:
```diff
@@ -0,0 +1,63 @@
+
+data_dirs <- list.dirs('data', recursive=FALSE)
+output_database <- "ref" # "test"
+required_file_names <- c("ref_circuit_details.csv", "ref_meta_data.json", "ref_raw_data.csv", "ref_site_details.csv")
```
In order to generate a set of reference data you'll need to run the tool at least once in order to get a representative sample of outputs, determine the pre event interval etc.
The reference data is based on the outputs in the first place, so I think it should be required for now. We could consider it once the tool is automated enough to generate those outputs, but I think the results should be manually inspected anyways so probably best to continue to require it.

phoebeheywood replied:
I just noticed that this script doesn't create a 'test_meta_data.json' to use when running the tool. Should it copy the ref file and save it with the test prefix?

exfletch replied:
The meta data should be the same for both the test and ref runs since we want to compare with the same settings.
The only context where it would need to be separate is when the test branch has added new settings, in which case copying the ref file probably wouldn't work.
Do you think I should change the name from `ref_meta_data.json` to `common_meta_data.json` or something to make it more obvious that the same settings should be used?

phoebeheywood replied:
I guess it's ok, but it does mean that the database file path is incorrect for the test. I also think that the metadata file is forwards compatible, and any missing entries are added in the next time it is saved, so I would tend towards copying the file

exfletch replied:
Good point. I'll copy the ref file as you've suggested, thanks!

phoebeheywood reviewed (commented):
The comparison works really well, I've left some comments mostly to do with improving the logging for better transparency

exfletch commented on validation/README.md:
```diff
@@ -0,0 +1,36 @@
+    1. Checkout reference version of tool (currently latest master branch)
+    2. Build database for the events in the `DER_disturbance_analysis/validation/data` directory using `build_validation_databases.R`
+    3. Run tool using metadata from the event directory as config json
+    4. Batch save results in the same directory as the data, entering `reference` as the file name
```
I think I've fixed this up

exfletch commented:
@phoebeheywood @amakarowsky I think I've addressed all the comments, could I get you to have a quick glance over to make sure I haven't missed anything before I merge?

amakarowsky commented:
LGTM
